transaction/core/src/amount/v2.rs

```diff
@@ -0,0 +1,341 @@
+    #[prost(message, required, tag = "1")]
+    pub commitment: CompressedCommitment,
+
+    /// `masked_value = value XOR_8 Blake2B(value_mask | shared_secret)`
+    #[prost(fixed64, required, tag = "2")]
+    pub masked_value: u64,
+
+    /// `masked_token_id = token_id XOR_8 Blake2B(token_id_mask |
+    /// shared_secret)` 8 bytes long when used, 0 bytes for older amounts
+    /// that don't have this.
+    #[prost(bytes, tag = "3")]
+    pub masked_token_id: Vec<u8>,
+}
+
+impl MaskedAmountV2 {
+    /// Creates a commitment `value*H + blinding*G`, and "masks" the commitment
+    /// secrets so that they can be recovered by the recipient.
+    ///
+    /// # Arguments
+    /// * `amount` - The amount information to be masked
+    /// * `shared_secret` - The shared secret, e.g. `rB` for transaction private
+    ///   key `r` and recipient public key `B`.
+    #[inline]
+    pub fn new(
+        amount: Amount,
+        tx_out_shared_secret: &RistrettoPublic,
+    ) -> Result<Self, AmountError> {
+        let amount_shared_secret = get_amount_shared_secret(tx_out_shared_secret);
+        let (value_mask, token_id_mask, blinding) = get_blinding_factors(&amount_shared_secret);
+
+        // Pedersen generators
+        let generator = generators(*amount.token_id);
+
+        // Pedersen commitment `v*H_i + b*G`.
+        let commitment = CompressedCommitment::new(amount.value, blinding, &generator);
+
+        // The value is XORed with the first 8 bytes of the mask.
+        let masked_value: u64 = amount.value ^ value_mask;
+
+        // The token_id is XORed with the first 8 bytes of the mask.
+        let masked_token_id_val: u64 = *amount.token_id ^ token_id_mask;
+        let masked_token_id = masked_token_id_val.to_le_bytes().to_vec();
+
+        Ok(Self {
+            commitment,
+            masked_value,
+            masked_token_id,
+        })
+    }
+
+    /// Returns the amount underlying the masked amount, given the shared
+    /// secret.
+    ///
+    /// Value is denominated in smallest representable units (e.g. "picoMOB").
+    ///
+    /// # Arguments
+    /// * `tx_out_shared_secret` - The shared secret, e.g. `rB`.
+    pub fn get_value(
+        &self,
+        tx_out_shared_secret: &RistrettoPublic,
+    ) -> Result<(Amount, Scalar), AmountError> {
+        let amount_shared_secret = get_amount_shared_secret(tx_out_shared_secret);
+        self.get_value_from_amount_shared_secret(&amount_shared_secret)
+    }
+
+    /// Returns the amount underlying the masked amount, given the amount shared
+    /// secret.
+    ///
+    /// Generally, the recipient knows the TxOut shared secret, and it's more
+    /// convenient to use the other API. This API allows that the sender or
+    /// recipient could selectively disclose the amount_shared_secret to a
+    /// third party who can then use this API to audit the value of
+    /// a TxOut without having permissions to do other things that require the
+    /// TxOut shared secret.
```

Comment:
:upside_down_face: 
notes on the suggestion:
* Prefer explicit over implicit, thus the function name
* I very well could be wrong, but I think of the entire module/struct as the API versus an individual `function`.
```suggestion
    /// Generally, the recipient knows the TxOut shared secret, and it's more
    /// convenient to use [get_value()]. This function allows that the sender or
    /// recipient could selectively disclose the `amount_shared_secret` to a
    /// third party who can then use this function to audit the value of
    /// a TxOut without having permissions to do other things that require the
    /// TxOut shared secret.
```